### PR TITLE
PGOV-840 Fix WSOD on certain search keywords

### PR DESCRIPTION
**Pull request by kxwang**

### CODE REVIEW

**Review by RichardDavies — COMMENTED** [review]
I have some concerns with the impact of how this modifies search queries so I did some additional testing and discovered some things that may be a problem.

File: web/modules/custom/portland/portland.module [hunk 1]
```diff
@@ -578,6 +578,28 @@ function portland_search_api_query_alter(\Drupal\search_api\Query\QueryInterface
...
+  // Must get the original keys
+  $keys = $query->getOriginalKeys();
+  // OR and AND in upper case are reseved keywords of Solr
+  $keys = strtolower($keys);
+  // "or" normally means "oregon"
+  $keys = preg_replace("/\bor\b/", "oregon", $keys);
+  // https://lucene.apache.org/core/8_7_0/queryparser/org/apache/lucene/queryparser/classic/package-summary.html#Escaping_Special_Characters
+  // Remove all Solr special characters
+  $keys = trim(preg_replace("/[\+\-\|\!\(\)\{\}\[\]\^\~\*\?\:\\\\\/]/", "", $keys));
+  // Allow query like "pay and contest"
```
[the remark this answers is not in the document]
Comment (kxwang) [reply web/modules/custom/portland/portland.module]: Removed this line.

File: web/modules/custom/portland/portland.module [hunk 2]
```diff
@@ -578,6 +578,28 @@ function portland_search_api_query_alter(\Drupal\search_api\Query\QueryInterface
...
+  $keys = str_replace("&", " and ", $keys);
```
[the remark this answers is not in the document]
Comment (kxwang) [reply web/modules/custom/portland/portland.module]: These are preserved now. Only trailing &'s are problematic.

**Comment by kxwang** [comment]
@RichardDavies Thanks for the rigorous review. After some further testing, I've revamped the logic and only replace characters in a few specific scenarios. Please review again. Thanks!

File: web/modules/custom/portland/portland.module [hunk 3]
```diff
@@ -585,22 +585,29 @@ function portland_search_api_solr_query_alter(\Solarium\QueryType\Select\Query\Q
...
+    // Convert "OR" and "||" at the start or end to "or"
+    "/^OR\b|\bOR$/",
+    "/^\|\|(?!\w)|(?!\w)\|\|$/",
+    // Convert "AND" and "&&" at the start or end to "and"
+    "/^AND\b|\bAND$/",
+    "/^\&\&(?!\w)|(?!\w)\&\&$/",
+    // "+" and "-" can't be at the end with space in any combination
+    "/\s[\+\- \"]+$/",
+    // "^" has to be appended immediately before a number
+    "/\^[^0-9\.]+/",
```
Comment (RichardDavies) [inline web/modules/custom/portland/portland.module:599]: FYI I still get an error if I search for "^1", "^0.1", or "^.1", although admittedly it's highly unlikely a user will search for those strings.

**Review by RichardDavies — APPROVED** [review]
This looks good to me now. All of my major concerns were addressed. I still have a couple of minor comments that may or may not warrant another tweak, so I'm approving it unless you want to tweak it further.

File: web/modules/custom/portland/portland.module [hunk 4]
```diff
@@ -585,22 +585,29 @@ function portland_search_api_solr_query_alter(\Solarium\QueryType\Select\Query\Q
...
-  $keys = preg_replace("/(\&+|^\|\|$)/", "", $keys);
+  $keys = preg_replace([
+    // Convert "OR" and "||" at the start or end to "or"
+    "/^OR\b|\bOR$/",
+    "/^\|\|(?!\w)|(?!\w)\|\|$/",
+    // Convert "AND" and "&&" at the start or end to "and"
+    "/^AND\b|\bAND$/",
+    "/^\&\&(?!\w)|(?!\w)\&\&$/",
+    // "+" and "-" can't be at the end with space in any combination
+    "/\s[\+\- \"]+$/",
```
[the remark this answers is not in the document]
Comment (kxwang) [reply web/modules/custom/portland/portland.module:597]: Add comments to include double quote and space.

File: web/modules/custom/portland/portland.module (hunk 3 above)
Comment (kxwang) [reply web/modules/custom/portland/portland.module:599]: Ah, good catch! Added a pattern to detect that.

**Review by RichardDavies — APPROVED** [review]

**Pull request merged**